src/app/EventManagement.cpp

```diff
@@ -590,46 +571,118 @@ CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWr
+                             .Cluster(aEvent.mClusterId)
+                             .Event(aEvent.mEventId)
+                             .EndOfEventPathIB()
+                             .GetError());
+
+    ReturnErrorOnFailure(builder.CreateErrorStatus().EncodeStatusIB(aStatus).GetError());
+
+    ReturnErrorOnFailure(builder.EndOfEventStatusIB().GetError());
+
+    ReturnErrorOnFailure(aWriter.EndContainer(containerType));
+    ReturnErrorOnFailure(aWriter.Finalize());
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EventManagement::CheckEventContext(EventLoadOutContext * eventLoadOutContext,
+                                              const EventManagement::EventEnvelopeContext & event)
 {
     if (eventLoadOutContext->mCurrentEventNumber < eventLoadOutContext->mStartingEventNumber)
     {
-        return false;
+        return CHIP_ERROR_UNEXPECTED_EVENT;
     }
 
-    if (event.mFabricIndex != kUndefinedFabricIndex && eventLoadOutContext->mFabricIndex != event.mFabricIndex)
+    if (event.mFabricIndex != kUndefinedFabricIndex && eventLoadOutContext->mSubjectDescriptor.fabricIndex != event.mFabricIndex)
     {
-        return false;
+        return CHIP_ERROR_UNEXPECTED_EVENT;
     }
 
     ConcreteEventPath path(event.mEndpointId, event.mClusterId, event.mEventId);
+    CHIP_ERROR ret = CHIP_ERROR_UNEXPECTED_EVENT;
+
+    bool isPathInterestedByConcretePath = false;
+
     for (auto * interestedPath = eventLoadOutContext->mpInterestedEventPaths; interestedPath != nullptr;
          interestedPath        = interestedPath->mpNext)
     {
         if (interestedPath->IsEventPathSupersetOf(path))
         {
-            return true;
+            ret = CHIP_NO_ERROR;
+            if (!interestedPath->HasEventWildcard())
+            {
+                isPathInterestedByConcretePath = true;
+                break;
+            }
         }
     }
-    return false;
+
+    ReturnErrorOnFailure(ret);
+
+    Access::RequestPath requestPath{ .cluster = event.mClusterId, .endpoint = event.mEndpointId };
+    Access::Privilege requestPrivilege = RequiredPrivilege::ForReadEvent(path);
+    CHIP_ERROR aclError                = CHIP_NO_ERROR;
+
+#if CONFIG_IM_BUILD_FOR_UNIT_TEST
+    switch (GetInstance().mBypassACL)
+    {
+    case BypassACL::kAlwaysPass:
+        aclError = CHIP_NO_ERROR;
+        break;
+    case BypassACL::kAlwaysFail:
+        aclError = CHIP_ERROR_ACCESS_DENIED;
+        break;
+    case BypassACL::kNoBypass:
+#endif
+        aclError = Access::GetAccessControl().Check(eventLoadOutContext->mSubjectDescriptor, requestPath, requestPrivilege);
+#if CONFIG_IM_BUILD_FOR_UNIT_TEST
+        break;
+    }
+#endif
+
+    if (aclError != CHIP_NO_ERROR)
+    {
+        ReturnErrorCodeIf(aclError != CHIP_ERROR_ACCESS_DENIED, aclError);
+        if (isPathInterestedByConcretePath)
+        {
+            ret = CHIP_ERROR_ACCESS_DENIED;
+        }
+        else
+        {
+            ret = CHIP_ERROR_UNEXPECTED_EVENT;
```

Comment:
Is this more like "unexpected error" or actually meaning that the event (being processed) is not expected?

Reply:
This means the event being processed is not interested